numpy/lib/arraysetops.py

```diff
@@ -593,6 +615,75 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
     # Ensure that iteration through object arrays yields size-1 arrays
     if ar2.dtype == object:
         ar2 = ar2.reshape(-1, 1)
+    # Convert booleans to uint8 so we can use the fast integer algorithm
+    if ar1.dtype == bool:
+        ar1 = ar1.view(np.uint8)
+    if ar2.dtype == bool:
+        ar2 = ar2.view(np.uint8)
+
+    # Check if we can use a fast integer algorithm:
+    integer_arrays = (np.issubdtype(ar1.dtype, np.integer) and
+                      np.issubdtype(ar2.dtype, np.integer))
+
+    if kind not in {None, 'mergesort', 'dictionary'}:
+        raise ValueError(
+            "Invalid kind: {0}. ".format(kind)
+            + "Please use None, 'mergesort' or 'dictionary'.")
+
+    if integer_arrays and kind in {None, 'dictionary'}:
+        ar2_min = np.min(ar2)
+        ar2_max = np.max(ar2)
+        ar1_size = ar1.size
+        ar2_size = ar2.size
+
+        # Check for integer overflow
+        with np.errstate(over='raise'):
+            try:
+                ar2_range = ar2_max - ar2_min
+
+                # Optimal performance is for approximately
+                # log10(size) > (log10(range) - 2.27) / 0.927.
+                # However, here we set the requirement that
+                # the intermediate array can only be 6x
+                # the combined memory allocation of the original
+                # arrays.
+                # (see discussion on 
+                # https://github.com/numpy/numpy/pull/12065)
+                below_memory_constraint = (
+                    ar2_range <= 6 * (ar1_size + ar2_size)
+                )
+            except FloatingPointError:
+                below_memory_constraint = False
+
+        # Use the fast integer algorithm
+        if below_memory_constraint or kind == 'dictionary':
```

Comment:
This doesn't seem right.  If `kind == 'dictionary'` you ignore the overflow path above?  It seems there is a test missing for, `kind="dictionary"` when the overflow path would be hit?

Reply:
Good catch. Indeed, if the overflow happens with `ar2_max-ar2_min`, then the `ar2 - ar2_min` will overflow as well. Will fix.

Reply:
I was briefly considering if you should use Python integers rather than the `errstate`.  Then you still need to check against `np.iinfo(arr.dtype).max` though I guess.  So I am not sure it is better.

Reply:
Added.